sound/soc/sof/sof-pci-dev.c

```diff
@@ -207,14 +207,13 @@ static int sof_pci_probe(struct pci_dev *pci,
 	mach = snd_soc_acpi_find_machine(desc->machines);
 #if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
 	if (!mach) {
-		dev_warn(dev, "No matching ASoC machine driver found - falling back to HDA codec\n");
-		mach = snd_soc_acpi_intel_hda_machines;
-		mach->sof_fw_filename = desc->nocodec_fw_filename;
-
-		/*
-		 * TODO: we need to find a way to check if codecs are actually
-		 * present
-		 */
+		struct hdac_bus bus;
+
+		if (hda_codec_detect(pci, &bus, FALSE) > 0) {
```

Comment:
that is still adding a new step and new API to the core that's HDaudio specific. You will also do things multiple times since the "bus" variable is local and all inits are gone once you leave the scope of this block.